FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc

```diff
@@ -0,0 +1,258 @@
+                  edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                  pool.runAsync(
+                      h2, [&count]() { ++count; }, errorContext);
+                }) |
+                lastTask(edm::WaitingTaskHolder(group, &waitTask));
+
+      auto h2 = first([&pool, &count](edm::WaitingTaskHolder h) {
+                  edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                  pool.runAsync(
+                      h2, [&count]() { ++count; }, errorContext);
+                }) |
+                lastTask(edm::WaitingTaskHolder(group, &waitTask));
+      h2.doneWaiting(std::exception_ptr());
+      h1.doneWaiting(std::exception_ptr());
+    }
+    waitTask.waitNoThrow();
+    REQUIRE(count.load() == 2);
+    REQUIRE(waitTask.done());
+    REQUIRE(not waitTask.exceptionPtr());
+  }
+
+  SECTION("Exceptions") {
+    SECTION("One async call") {
+      std::atomic<int> count{0};
+
+      oneapi::tbb::task_group group;
+      edm::FinalWaitingTask waitTask{group};
+
+      {
+        using namespace edm::waiting_task::chain;
+        auto h = first([&pool, &count](edm::WaitingTaskHolder h) {
+                   edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                   pool.runAsync(
+                       std::move(h2), [&count]() { throw std::runtime_error("error"); }, errorContext);
+                 }) |
+                 lastTask(edm::WaitingTaskHolder(group, &waitTask));
+        h.doneWaiting(std::exception_ptr());
+      }
+      REQUIRE_THROWS_WITH(
+          waitTask.wait(),
+          Catch::Contains("error") and Catch::Contains("StdException") and Catch::Contains("WaitingThreadPool test"));
+      REQUIRE(count.load() == 0);
+    }
+
+    SECTION("Two async calls") {
+      std::atomic<int> count{0};
+
+      oneapi::tbb::task_group group;
+      edm::FinalWaitingTask waitTask{group};
+
+      {
+        using namespace edm::waiting_task::chain;
+        auto h = first([&pool, &count](edm::WaitingTaskHolder h) {
+                   edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                   pool.runAsync(
+                       h2,
+                       [&count]() {
+                         if (count.load() == 1) {
+                           throw cms::Exception("error 1");
+                         }
+                         ++count;
+                       },
+                       errorContext);
+                   pool.runAsync(
+                       h2,
+                       [&count]() {
+                         if (count.load() == 1) {
+                           throw cms::Exception("error 2");
+                         }
+                         ++count;
+                       },
+                       errorContext);
+                 }) |
+                 lastTask(edm::WaitingTaskHolder(group, &waitTask));
+        h.doneWaiting(std::exception_ptr());
+      }
+      REQUIRE_THROWS_AS(waitTask.wait(), cms::Exception);
+      REQUIRE(count.load() == 1);
```

Comment:
Is this actually guaranteed to work? Couldn't `"1"` thread start and pause at line 134 then `"2"` thread finishes and only after that `"1"` continues? In that case `count.load() == 2`.

Reply:
Good point, thanks. I need to use `fetch_add()`.